src/libertem/io/dataset/hdf5.py

```diff
@@ -266,23 +333,81 @@ def _preprocess(self, tile_data, tile_slice):
+        source_tile_slice = Slice(shape=source_shape, origin=source_origin)
+        dest_origin = ([0] * len(buf_shape))
+        dest_shape = source_shape
+        dest_tile_slice = Slice(shape=dest_shape, origin=dest_origin)
+        offset_origin = (o[0] + 1,) + (0, 0, 0)
+        offset_shape = Shape((1,) + (self.sync_offset,) + (s[2], s[3]), sig_dims=2)
+        offset_tile_slice = Slice(shape=offset_shape, origin=offset_origin)
+        dest_offset_origin = (0,) + (s[1] - self.sync_offset,) + (0, 0)
+        dest_offset_shape = offset_shape
+        dest_offset_tile_slice = Slice(shape=dest_offset_shape, origin=dest_offset_origin)
+        return source_tile_slice, dest_tile_slice, offset_tile_slice, dest_offset_tile_slice
+
     def _get_tiles_normal(self, tiling_scheme, tileshape_nd, dest_dtype="float32"):
-        data = np.ndarray(tileshape_nd, dtype=dest_dtype)
+        reshaped_tileshape_nd = tileshape_nd
+        if self.do_reshape:
+            reshaped_tileshape_nd = Shape(tileshape_nd, sig_dims=2).flatten_nav()
+        data = np.ndarray(reshaped_tileshape_nd, dtype=dest_dtype)
         with self.reader.get_h5ds() as dataset:
             subslices = self._get_subslices(
                 tiling_scheme=tiling_scheme,
                 tileshape_nd=tileshape_nd,
             )
             for scheme_idx, tile_slice in subslices:
-                tile_slice_flat = tile_slice.flatten_nav(self.meta.shape)
+                if self.do_reshape:
+                    tile_slice = self._get_reshaped_tile_slice(tile_slice)
                 if tuple(tile_slice.shape) != tuple(tileshape_nd):
                     # at the border, can't reuse buffer
                     border_data = np.ndarray(tile_slice.shape, dtype=dest_dtype)
                     buf = border_data
                 else:
                     # reuse buffer
                     buf = data
-                dataset.read_direct(buf, source_sel=tile_slice.get())
+                source_tile_slice = tile_slice
+                buf_shape = Shape(buf.shape, sig_dims=2)
+                dest_tile_slice = Slice(shape=buf_shape, origin=[0] * len(buf.shape))
+                if self.sync_offset != 0:
+                    if self.sync_offset > 0:
+                        slices_with_offset = self._get_tile_slice_with_positive_offset(
+                            tile_slice, buf.shape
+                        )
+                    # else:
+                    #    slices_with_offset = self._get_tile_slice_with_negative_offset(tile_slice)
+                    source_tile_slice = slices_with_offset[0]
+                    dest_tile_slice = slices_with_offset[1]
+                    offset_tile_slice = slices_with_offset[2]
+                    dest_offset_tile_slice = slices_with_offset[3]
+                dataset.read_direct(
+                    buf, source_sel=source_tile_slice.get(), dest_sel=dest_tile_slice.get()
```

Comment:
Why is `dest_sel` needed here, shouldn't that always be the whole `buf`? Also, what is the idea of doing two independent `read_direct` calls?

Reply:
The idea is to fill the `buf` in two steps: `(slice_0_origin_with_offset) -> (slice_0_end)` and `(slice_0_end) -> (slice_1_origin_with_offset)` and so on. Two `read_direct` calls were needed because I needed slices from two rows at once. I can also combine the slices and read once using `np.r_` or use numpy's `stride_tricks` if required. I'll definitely benchmark and try the different approaches.

> shouldn't that always be the whole buf?

I'm not sure, I'll look into it again.

Reply:
Ah, I think I know what you mean - it gets a bit complicated to fold the 1D offset into the ND slice that is then used to index the HDF5 dataset... then this approach may be needed.

> I'll definitely benchmark and try the different approaches.

:+1: